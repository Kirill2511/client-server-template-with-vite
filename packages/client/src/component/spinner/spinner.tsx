import React, { ElementType, forwardRef } from 'react';
import classNames from 'classnames';

export interface SpinnerProps {
  className?: string;
  color?: string;
  component?: string | ElementType;
  size?: 'sm';
  variant?: 'border' | 'grow';
  visuallyHiddenLabel?: string;
}

type Ref = HTMLDivElement | HTMLSpanElement;

export const Spinner = forwardRef<Ref, SpinnerProps>(
  (
    {
      className,
      color,
      component: Component = 'div',
      size,
      variant = 'border',
      visuallyHiddenLabel = 'Loading...',
      ...rest
    },
    ref,
  ) => {
    const _className = classNames(
      `spinner-${variant}`,
      `text-${color}`,
      size && `spinner-${variant}-${size}`,
      className,
    );

    return (
      <Component className={_className} role="status" {...rest} ref={ref}>
        <span className="visually-hidden">{visuallyHiddenLabel}</span>
      </Component>
    );
  },
);

Spinner.displayName = 'Spinner';
