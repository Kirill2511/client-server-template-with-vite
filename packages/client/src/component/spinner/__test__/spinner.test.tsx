import * as React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Spinner } from '../spinner';

test('loads and displays Spinner component', async () => {
  const { container } = render(<Spinner />);
  expect(container).toMatchSnapshot();
});

test('Spinner customize', async () => {
  const { container } = render(<Spinner className="bazinga" color="warning" component="h3" size="sm" variant="grow" />);
  expect(container).toMatchSnapshot();
  expect(container.firstChild).toHaveClass('bazinga');
  expect(container.firstChild).toHaveClass('spinner-grow');
  expect(container.firstChild).toHaveClass('text-warning');
  expect(container.firstChild).toHaveClass('spinner-grow-sm');
});
