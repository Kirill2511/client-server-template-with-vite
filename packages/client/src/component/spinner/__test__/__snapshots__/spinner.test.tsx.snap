// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner customize 1`] = `
<div>
  <h3
    class="spinner-grow text-warning spinner-grow-sm bazinga"
    role="status"
  >
    <span
      class="visually-hidden"
    >
      Loading...
    </span>
  </h3>
</div>
`;

exports[`loads and displays Spinner component 1`] = `
<div>
  <div
    class="spinner-border text-undefined"
    role="status"
  >
    <span
      class="visually-hidden"
    >
      Loading...
    </span>
  </div>
</div>
`;
